### PR TITLE
*: refine 'tidb_snapshot' behavior

### LINKED ISSUE #34529: Dumpling snapshot failed
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. create a table, insert some data
```
mysql> create database t;
Query OK, 0 rows affected (0.08 sec)
mysql> create table t.t1 (a int);
Query OK, 0 rows affected (0.08 sec)
mysql> insert into t.t1 values (1);
Query OK, 1 row affected (0.01 sec)
```
2. get current time
```
mysql> select now();
+---------------------+
| now()               |
+---------------------+
| 2022-05-10 06:20:46 |
+---------------------+
1 row in set (0.00 sec)
```
3.  drop database
mysql> drop database t;
Query OK, 0 rows affected (0.20 sec)

4. Use dumpling to export the deleted database, specifying parameter --snapshot "2022-05-10 06:20:46" => dumpling failed with error "Unknown database 't' "
```
/dumpling  -h "src-tidb-tidb.fb-br-debug-p65cb" -P "4000" -u "root" --output "s3://tmp/ttt?access-key=minioadmin&secret-access-key=minioadmin&endpoint=http%3a%2f%[2fminio.pingcap.net](http://2fminio.pingcap.net/)%3a9000&force-path-style=true" --filetype "csv" --filesize "256M" --rows "200" --threads "20" -B t --snapshot "2022-05-10 06:20:46"
[2022/05/10 06:21:28.133 +00:00] [INFO] [conn.go:45] ["cannot execute query"] [retryTime=1] [sql="SHOW CREATE DATABASE `t`"] [args=null] [error="sql: SHOW CREATE DATABASE `t`, args: []: Error 1049: Unknown database 't'"]
```
Notes:
1. This issue is only seem in recent nightly build ( This case was passed in 05/04 CI )
2. After dumpling snapshot failed, I can actually see the database if I connect to tidb and set @@tidb_snapshot="2022-05-10 06:20:46";

```
mysql> set @@tidb_snapshot="2022-05-10 06:20:46";
Query OK, 0 rows affected (0.00 sec)
mysql> show databases;
+--------------------+
| Database           |
+--------------------+
| INFORMATION_SCHEMA |
| METRICS_SCHEMA     |
| PERFORMANCE_SCHEMA |
| mysql              |
| t                  |
+--------------------+
5 rows in set (0.00 sec)
```
### 2. What did you expect to see? (Required)
Dumpling snapshot should be successful.
### 3. What did you see instead (Required)
dumpling failed with error "Unknown database 't' "
```
/dumpling  -h "src-tidb-tidb.fb-br-debug-p65cb" -P "4000" -u "root" --output "s3://tmp/ttt?access-key=minioadmin&secret-access-key=minioadmin&endpoint=http%3a%2f%[2fminio.pingcap.net](http://2fminio.pingcap.net/)%3a9000&force-path-style=true" --filetype "csv" --filesize "256M" --rows "200" --threads "20" -B t --snapshot "2022-05-10 06:20:46"
[2022/05/10 06:21:28.133 +00:00] [INFO] [conn.go:45] ["cannot execute query"] [retryTime=1] [sql="SHOW CREATE DATABASE `t`"] [args=null] [error="sql: SHOW CREATE DATABASE `t`, args: []: Error 1049: Unknown database 't'"]
```

### 4. What is your TiDB version? (Required)
/ # /tidb-server -V
Release Version: v6.1.0-nightly
Edition: Community
Git Commit Hash: 20ecaef36d747a9993fffaf04903e8f7c9dbf527
Git Branch: heads/refs/tags/v6.1.0-nightly
UTC Build Time: 2022-05-09 11:03:05
GoVersion: go1.18
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #34529

### What is changed and how it works?

This PR refines some behaviors for `tidb_snapshot` , that is:

1. To be compatible with some ecosystem tools, `tidb_snapshot` can be set/unset when in/out transaction.
2. when `tidb_snapshot` is set, the 'SELECT' will always use the infoschema and data according to the `tidb_snapshot`.
3. when `tidb_snapshot` is unset, the session should no longer use the `tidb_snapshot`.
4. 'SELECT ... FOR UPDATE ..' is not allowed when `tidb_snapshot` is set  in a pessimistic transaction 

Also fixed the bug #34529. After executing `ExecRestrictedSQL` with options `ExecOptionWithSnapshot` and `ExecOptionUseCurSession` , the snapshot info will be reset that will cause some unexpected error. So this PR fixed it by restoring the original settings.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
txn: 'SELECT ... FOR UPDATE ..' is not allowed when `tidb_snapshot` is set  in a pessimistic transaction 

```

### CODE REVIEW

File: sessiontxn/legacy/provider.go [hunk 1]
```diff
@@ -72,7 +72,7 @@ func (p *SimpleTxnContextProvider) OnInitialize(ctx context.Context, tp sessiont
 	switch tp {
 	case sessiontxn.EnterNewTxnDefault, sessiontxn.EnterNewTxnWithBeginStmt:
 		txnCtx := sessVars.TxnCtx
-		if tp != sessiontxn.EnterNewTxnWithBeginStmt || txnCtx.History != nil || txnCtx.IsStaleness {
+		if tp != sessiontxn.EnterNewTxnWithBeginStmt || txnCtx.History != nil || txnCtx.IsStaleness || sessVars.SnapshotTS > 0 {
```
Comment (cfzjywxk) [inline sessiontxn/legacy/provider.go:75]: Better to add some comments here as it's a bit complex by now.

File: executor/executor_test.go [hunk 2]
```diff
@@ -5357,6 +5358,122 @@ func TestHistoryRead(t *testing.T) {
...
+
+				init(isolation, setSnapshotBeforeTxn)
+				// When tidb_snapshot is set, should use the snapshot info schema
+				tk.MustQuery("show tables like 'his_%'").Check(testkit.Rows("his_t0"))
+
+				// When tidb_snapshot is set, select should use select ts
+				tk.MustQuery("select * from his_t0").Check(testkit.Rows("1 10"))
+				tk.MustQuery("select * from his_t0 where id=1").Check(testkit.Rows("1 10"))
+
+				// When tidb_snapshot is set, select ... for update should not be effected
```
Comment (cfzjywxk) [inline executor/executor_test.go:5434]: Does here mean `should be affected`? It's also needed to cover the `update/delete` paths which may do for update read first. 

The expectations seem to be not clear when `tidb_snapshot` is used together with pessimistic transaction semantics. /cc @MyonKeminta @you06

Comment (lcwangchao) [reply executor/executor_test.go:5434]: I think it can be "select ... for update should not be allowed"

File: sessiontxn/legacy/provider.go (hunk 1 above)
Comment (lcwangchao) [reply sessiontxn/legacy/provider.go:75]: done

**Review by tiancaiamao — APPROVED** [review]

**Review by cfzjywxk — APPROVED** [review]

**Pull request merged**